Posts/2021/03/DarwinJS-Contribution-LightningFastandEasyProvisioningofGitwithSSHKeyAuthenticationonWindows.md

```diff
@@ -0,0 +1,117 @@
+
+<!-- more -->
+
+## Reasons For Moving to SSH
+
+There are multiple reasons you may want to move your Windows developers to SSH authentication for Git:
+
+1. You want to get away from git storing local passwords - whether in the git config or in Windows Credentials (with the windows credential helper) because it is pure pain to walk people through how to find and update this password when they change it on the Git server.
+2. You want to avoid both http passwords and the http protocol for git.
+
+## Conventional Wisdom on SSH Configuration
+
+The conventional wisdom solution offers many steps that are roughly:
+
+1. Installing git manually.
+2. Installing the well known Windows SSH client Putty
+3. Installing Putty's key generator.
+4. Converting the non-compatible putty generated key into an ssh compatible one.
+5. Precisely placing the SSH key on disk.
+6. Precisely and ~~manually~~ permissioning the SSH key and it's parent folder (ssh is purposely fussy about this in order to keep the key secure).
+
+> Most of this can be avoided by simply using the full SSH client that is embedded inside of the Windows git client install.
+
+## The Cleanest Way (With Working Automation Code) 
+
+Besides the above pure pain, here are the additional things solved for in this code:
+
+1. Automatically installs Git - but only if necessary (idempotent)
+2. Automatically installs chocolatey to install Git - but only if necessary (idempotent)
+3. Automatically generates an SSH key - but only if necessary (idempotent) (which avoids killing a key that might be in use)
+4. Uses the Git's built-in SSH client to create SSH keys (avoids the complexity of the above conventional wisdom)
+5. Copies the public key to the clip board and pauses for the user to add it to the Git server (in their profile)
+6. Optionally does a SSH login test if you provide a value for: $SSHEndPointToGitForTesting
+
+## Solution Details
+
+This code can be run directly from GitLab with this command:
+
+```bash
+Invoke-Expression -command "Invoke-WebRequest -uri 'https://gitlab.com/missionimpossiblecode/MissionImpossibleCode/-/raw/master/install-gitwithssh.ps1' -UseBasicParsing -OutFile ./install-gitwithssh.ps1" ; . ./install-gitwithssh.ps1
+```
+
+If you want to download dynamically, but also want the test and instructions to work, then set these environment variables before calling the above:
+
+```bash
+$env:YourGitServerhttpURL="https://gitlab.com"
+$env:GitSSHUserAndEndPointForTesting="git@gitlab.com" #some Git servers might want the windows userid "git", which is specified as $env:username
+```
+
+You can also simply copy the code, hardcode the two variables and distribute it in your organization.
+
+## Main Code
+
+```bash
+
+# Set environment variables before calling in order to test
```

Comment:
Remove the blank line and change the label.

```suggestion
```powershell
# Set environment variables before calling in order to test
```